src/tools/miri/tests/pass/function_pointers.rs

```diff
@@ -80,9 +80,8 @@ fn main() {
     // but Miri currently uses a fixed address for monomorphic functions.
     assert!(return_fn_ptr(i) == i);
     assert!(return_fn_ptr(i) as unsafe fn() -> i32 == i as fn() -> i32 as unsafe fn() -> i32);
-    // We don't check anything for `f`. Miri gives it many different addresses
-    // but mir-opts can turn them into the same address.
-    let _val = return_fn_ptr(f) != f;
+    // Miri gives it many different addresses to different reifications of a generic function.
+    assert!(return_fn_ptr(f) != f);
```

Comment:
I put back the check demanded in https://github.com/rust-lang/rust/pull/109597/files/901be42bc09bde57ad9b19a3c2c432fd3482279c#r1339553254

Reply:
Why didn't you have to add a black_box to make that work?

Reply:
I propose we consider the Reify* casts (both fn and closure) as non deterministic kn this pass too. Instead of trying to represent them, I assign them a new opaque VnIndex each time. See the change in the same commit as the test.

Reply:
Ah I see, makes sense. Not sure if it's strictly needed but it seems like a good conservative choice for now.